tensorboard/loader.py

```diff
@@ -0,0 +1,414 @@
+
+  def __init__(self, path, start_offset=0):
+    """Creates new instance.
+
+    Args:
+      path: Path of file. This can be on a remote file system if the
+          TensorFlow build supports it.
+      start_offset: Byte offset to seek in file once it's opened.
+
+    :type path: str
+    :type start_offset: int
+    """
+    self.path = tf.compat.as_text(path)
+    self._offset = start_offset
+    self._reader = None  # type: tf.pywrap_tensorflow.PyRecordReader
+    self._is_closed = False
+
+  def get_next_record(self):
+    """Reads record from file.
+
+    Returns:
+      A Record or None if no more were available.
+
+    Raises:
+      IOError: On open or read error, or if close was called.
+
+    :rtype: Record
+    """
+    if self._is_closed:
+      raise IOError('%s is closed' % self)
+    if self._reader is None:
+      self._reader = self._open()
+    try:
+      with tf.errors.raise_exception_on_not_ok_status() as status:
+        self._reader.GetNext(status)
+    except (tf.errors.DataLossError, tf.errors.OutOfRangeError):
+      # We ignore partial read exceptions, because a record may be truncated.
+      # PyRecordReader holds the offset prior to the failed read, so retrying
+      # will succeed.
+      return None
+    return Record(self._reader.record(), self._reader.offset())
+
+  def get_offset(self):
+    """Returns current offset in file.
+
+    Raises:
+      IOError: If close was called.
+
+    :rtype: int
+    """
+    if self._is_closed:
+      raise IOError('%s is closed' % self)
+    if self._reader is None:
+      return self._offset
+    return self._reader.offset()
+
+  def get_size(self):
+    """Returns byte length of file.
+
+    This is guaranteed to return a number greater than or equal to the
+    offset of the last record returned by get_next_record().
+
+    This method can be called after the instance has been closed.
+
+    Raises:
+      IOError: If file has shrunk from last offset or start offset, or
+          this instance is closed.
+
+    :rtype: int
+    """
+    length = tf.gfile.Stat(self.path).length
+    if length < self.get_offset():
+      raise IOError('File shrunk %d < %d: %s' %
+                    (length, self._offset, self.path))
+    return length
+
+  def close(self):
+    """Closes record reader if open.
+
+    Further reads are not permitted after this method is called.
+    """
+    if self._is_closed:
+      return
+    if self._reader is not None:
+      self._reader.Close()
+    self._is_closed = True
+    self._reader = None
+
+  def _open(self):
+    with tf.errors.raise_exception_on_not_ok_status() as status:
+      return tf.pywrap_tensorflow.PyRecordReader_New(
+          tf.resource_loader.readahead_file_path(tf.compat.as_bytes(self.path)),
+          self._offset, '', status)
+
+  def __str__(self):
+    return u'RecordReader{%s}' % self.path
+
+
+@util.closeable
+@six.python_2_unicode_compatible
+class BufferedRecordReader(object):
```

Comment:
Eh, I like the current name because `Buffered.*Reader` is a common pattern and it draws analogy to the original [`BufferedReader`](https://docs.oracle.com/javase/7/docs/api/java/io/BufferedReader.html), which immediately clarifies its purpose.